lfs.c

```diff
@@ -2390,11 +2438,15 @@ int lfs_file_opencfg(lfs_t *lfs, lfs_file_t *file,
 
     // deorphan if we haven't yet, needed at most once after poweron
     if ((flags & 3) != LFS_O_RDONLY) {
+#ifdef LFS_READONLY
+        return LFS_ERR_ROFS;
+#else
```

Comment:
I would think that files can only ever be opened in `LFS_O_RDONLY` mode, so this entire if-block can just be excluded with the #if .

Reply:
There's also another similar block down below in this function which could also be #if'd out.

Reply:
Actually that's a good point, an `LFS_ASSERT((flags & 3) == LFS_O_RDONLY)` would work better here.

Reply:
I am not a big fan of "crashing" with an assert, instead of returning an error when a `lfs_file_opencfg()` is called with the wrong arguments, but I'll implement it that way, if you want.

Let's say an OS tries to write to a file mounted using the LittleFS filesystem, but LittleFS is compiled for read-only. Wouldn't it be more elegant to return an error indicating it's not possible to write, instead of asserting ?

Reply:
At the moment, it should be an assert for consistency (writing a file opened as rdonly currently asserts, for example).

Returning an error can be more graceful if there is an OS that understands that error, but for many littlefs applications this isn't the case. It's easy for the context of the error to get lost as "file open failed". Looking at POSIX errors, which littlefs follows, writing to an rdonly file only returns EINVAL. An assert takes you directly to the source of the problem, sort of like a poor-man's stack trace.

There's also a code cost for returning/handling errors, an assert can be dropped at compile time. Runtime checks can be added in the OS, but they can't be removed. This makes asserts preferred for conditions that should never happen unless there is a programmer mistake.

But you're not the first person to comment on this, maybe it would be worth adding a configuration option to change the use-error asserts to return errors. The next steps there would be to create a separate issue/PR.

Reply:
Alright, thanks for the detailed explanation. I can understand the reasoning indeed.

I have just pushed a new commit implementing the `LFS_ASSERT` for write attempts, as well as more code commented out, in case of `LFS_READONLY`.

Reply:
One idea would be to also #if out the write-based constants, like `LFS_O_WRONLY`.  Then it would be impossible to compile code that opened files for writing.

Reply:
Oh, that's an even better idea! @dpgeorge with the great ideas

Reply:
Okay, I just pushed another commit implementing this idea.
I had to make some extra `#ifdef` sections in other functions checking for these flags, so please carefully review that.
What I also did, is marking each `#endif` section with a comment indicating which `#ifdef` it is corresponding to.